WordPressAuthenticator/OAuth/ProofKeyForCodeExchange.swift

```diff
@@ -25,21 +25,92 @@ struct ProofKeyForCodeExchange {
         }
     }
 
-    let codeVerifier: String
+    let codeVerifier: CodeVerifier
     let method: Method
 
-    init(codeVerifier: String, method: Method) {
+    init(codeVerifier: CodeVerifier = .makeRandomCodeVerifier(), method: Method = .s256) {
         self.codeVerifier = codeVerifier
         self.method = method
     }
 
     var codeCallenge: String {
         switch method {
         case .s256:
-            // TODO: code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
-            fatalError()
+            // The spec defines code_challenge for the s256 mode as:
+            //
+            // code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
+            //
+            // We don't need the ASCII conversion, because we build `CodeVerifier` from URL safe
+            // characters.
+            let rawData = Data(codeVerifier.rawValue.utf8)
+            let hashedData: Data = rawData.sha256Hashed()
+            return hashedData.base64URLEncodedString()
         case .plain:
-            return codeVerifier
+            return codeVerifier.rawValue
         }
     }
 }
+
+extension ProofKeyForCodeExchange {
+
+    struct CodeVerifier: Equatable {
+
+        let rawValue: String
+
+        static let allowedCharacters = Character.urlSafeCharacters
+        static let minimumLength = 43
+        static let maximumLength = 128
```

Comment:
These two constants could be combined into one: `static let acceptableLength = 43...128`?

Reply:
Good point. Addressed in 93f584d6